spec/classes/plugin/nginx_spec.rb

```diff
@@ -0,0 +1,39 @@
+require 'spec_helper'
+
+describe 'letsencrypt::plugin::nginx' do
+  on_supported_os.each do |os, os_facts|
+    context "on #{os}" do
+      let(:facts) { os_facts }
+      let(:params) { {} }
+      let(:pre_condition) do
+        <<-PUPPET
+        class { 'letsencrypt':
+          email => 'foo@example.com',
+        }
+        PUPPET
+      end
+
+      context 'with default parameters' do
+        it { is_expected.to compile.with_all_deps }
+
+        let(:package_name) do
+          if os_facts[:os]['family'] == 'RedHat' && os_facts[:os]['release']['major'] == '7'
```

Comment:
It is valid to refer to `facts` here. Then you use `let(:facts) { ...}`. So it will also take any overrides into account. That can be another reason to prefer `let`.

Reply:
Please excuse the probably stupid question, but why use `let(:facts) { ... }` in the first place. What is `let(:...) { ... }` good for? I can't read the complete rspec docs in 15 minutes, but maybe a chapter ;).

Reply:
Thank you! I think I understood. :)

Reply:
Now this makes some sense:
```puppet
describe 'with manage_package => false' do
          let(:params) { super().merge(manage_package: false, package_name: 'nginx-package') }

          it { is_expected.not_to contain_package('nginx-package') }
        end
```

Reply:
Should be improved in one of the commits.